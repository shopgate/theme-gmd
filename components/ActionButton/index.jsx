/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import IndicatorCircle from 'Components/IndicatorCircle';
import RippleButton from 'Components/RippleButton';
import styles from './style';

/**
 * The action button component.
 */
class ActionButton extends Component {
  static propTypes = {
    ...RippleButton.propTypes,
    onClick: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    noGap: PropTypes.bool,
  };

  static defaultProps = {
    ...RippleButton.defaultProps,
    loading: false,
    type: 'primary',
    flat: true,
    noGap: false,
  };

  static clickDelay = 300;

  /**
   * The click handler
   * @param {Object} event The event object for the click handler
   */
  handleClick = (event) => {
    const { clickDelay } = this.constructor;

    setTimeout(() => {
      this.props.onClick(event);
    }, clickDelay);
  };

  /**
   * Getter for the calculated button props.
   * @returns {Object}
   */
  get buttonProps() {
    const buttonProps = {
      className: this.props.className,
      disabled: this.props.disabled,
      flat: this.props.flat,
      type: this.props.type,
    };

    return buttonProps;
  }

  /**
   * The render function.
   * @returns {JSX}
   */
  render() {
    const containerClass = this.props.noGap ? styles.noGapContainer : styles.container;

    if (this.props.loading) {
      return (
        <div className={styles.containerCircle}>
          <IndicatorCircle />
        </div>
      );
    }

    return (
      <div className={containerClass}>
        <RippleButton {...this.buttonProps} onClick={this.handleClick}>
          {this.props.children}
        </RippleButton>
      </div>
    );
  }
}

export default ActionButton;
