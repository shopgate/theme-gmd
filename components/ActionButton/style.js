/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';
import variables from 'Styles/variables';

const halfGapBig = variables.gap.big / 2;

const containerBase = {
  textAlign: 'center',
};

const container = css({
  ...containerBase,
  margin: `${halfGapBig}px 0`,
}).toString();

const noGapContainer = css({
  ...containerBase,
}).toString();

const containerCircle = css({
  ...containerBase,
  margin: `${halfGapBig + 4}px 0`,
}).toString();

export default {
  container,
  containerCircle,
  noGapContainer,
};
