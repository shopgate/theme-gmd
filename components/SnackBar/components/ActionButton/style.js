/**
 *  Copyright (c) 2017 - present, Shopgate, Inc. All rights reserved.
 *
 *  This source code is licensed under the Apache 2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 */
import { css } from 'glamor';
import colors from 'Styles/colors';

const actionButton = css({
  color: colors.primary,
  flexGrow: 1,
  textTransform: 'uppercase',
  textAlign: 'right',
  padding: '7px 0',
  wordBreak: 'keep-all',
  ':focus': {
    outline: 'none',
  },
}).toString();

export default {
  actionButton,
};
