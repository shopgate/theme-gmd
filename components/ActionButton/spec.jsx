/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import IndicatorCircle from 'Components/IndicatorCircle';
import RippleButton from 'Components/RippleButton';
import ActionButton from './index';

describe('<ActionButton />', () => {
  let renderedElement;
  let mockOnClick;

  jest.useFakeTimers();

  /**
   * Renders the component.
   * @param {Object} props The component props.
   */
  const renderComponent = (props = {}) => {
    renderedElement = shallow((
      <ActionButton {...props}>
        Action Button
      </ActionButton>
    ));
  };

  describe('Given the component was mounted to the DOM', () => {
    beforeEach(() => {
      mockOnClick = jest.fn();

      renderComponent({
        onClick: mockOnClick,
      });
    });

    it('should match snapshot', () => {
      expect(renderedElement).toMatchSnapshot();
    });

    it('should not show the loading indicator by default', () => {
      const indicator = renderedElement.find(IndicatorCircle);
      expect(indicator.length).toBe(0);
    });

    describe('Given the loading prop is set to true', () => {
      beforeEach(() => {
        renderedElement.setProps({ loading: true });
      });

      it('should match snapshot', () => {
        expect(renderedElement).toMatchSnapshot();
      });

      it('should show the loading indicator', () => {
        const indicator = renderedElement.find(IndicatorCircle);
        expect(indicator.length).toBe(1);
      });
    });

    describe('Given the component gets clicked', () => {
      beforeEach(() => {
        renderedElement.find(RippleButton).simulate('click');
      });

      it('should use setTimeout for delaying the onClick handler', () => {
        expect(setTimeout.mock.calls.length).toBe(1);
        expect(setTimeout.mock.calls[0][1]).toBe(ActionButton.clickDelay);
      });

      it('should eventually call the onClick handler', () => {
        jest.runOnlyPendingTimers();

        expect(mockOnClick).toBeCalled();
        expect(mockOnClick.mock.calls.length).toBe(1);
      });
    });
  });
});
