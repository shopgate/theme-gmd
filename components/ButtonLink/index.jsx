/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ParsedLink from '@shopgate/pwa-common/components/Router/helpers/parsed-link';
import ActionButton from 'Components/ActionButton';

/**
 * Simple wrapper around ActionButton so it's easy to render buttons which behave as links.
 */
class ButtonLink extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    noGap: PropTypes.bool,
  };

  static defaultProps = {
    noGap: false,
  };

  /**
   * Prepares the link.
   * @param {Object} props Props
   */
  constructor(props) {
    super(props);
    this.link = new ParsedLink(this.props.href);
  }

  /**
   * Opens the given url on click.
   * @returns {undefined}
   */
  handleClick = () => this.link.open();

  /**
   * Renders an ActionButton and handles link handling.
   * @returns {XML}
   */
  render() {
    return (
      <ActionButton onClick={this.handleClick} type="secondary" noGap={this.props.noGap}>
        {this.props.children}
      </ActionButton>
    );
  }
}

export default ButtonLink;
