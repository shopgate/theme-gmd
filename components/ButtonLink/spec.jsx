/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { mount } from 'enzyme';
import ParsedMockLink from '@shopgate/pwa-common/components/Router/helpers/parsed-link/index.mock';
import ActionButton from '../ActionButton';
import ButtonLink from './index';

const mockedParsedLink = new ParsedMockLink('http://example.com');
jest.mock(
  '@shopgate/pwa-common/components/Router/helpers/parsed-link',
  () => () => mockedParsedLink
);

describe('<ButtonLink>', () => {
  describe('On click action', () => {
    beforeAll(() => {
      ActionButton.clickDelay = 0;
    });
    it('should create component and open page on click', (done) => {
      const component = mount(
        <ButtonLink href="https://example.com">
          Text inside
        </ButtonLink>
      );
      expect(component).toMatchSnapshot();
      expect(mockedParsedLink.openFunctionMock.mock.calls.length).toBe(0);
      component.find('RippleButton').simulate('click');
      component.update();
      setTimeout(() => {
        expect(mockedParsedLink.openFunctionMock.mock.calls.length).toBe(1);
        done();
      }, ActionButton.clickDelay + 1);
    });
  });
});
