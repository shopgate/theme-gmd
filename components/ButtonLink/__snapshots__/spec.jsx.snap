// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonLink> On click action should create component and open page on click 1`] = `
<ButtonLink
  href="https://example.com"
  noGap={false}
>
  <ActionButton
    className=""
    disabled={false}
    flat={true}
    loading={false}
    noGap={false}
    onClick={[Function]}
    rippleClassName=""
    rippleSize={null}
    type="secondary"
    wrapContent={true}
  >
    <div
      className="css-omcrdt"
    >
      <RippleButton
        className=""
        disabled={false}
        flat={true}
        onClick={[Function]}
        rippleClassName=""
        rippleSize={null}
        type="secondary"
        wrapContent={true}
      >
        <Button
          className=""
          disabled={false}
          flat={true}
          onClick={[Function]}
          type="secondary"
          wrapContent={false}
        >
          <Button
            className="css-20ad7t "
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="css-20ad7t  css-1rn65c0"
              data-test-id="Button"
              disabled={false}
              onClick={[Function]}
            >
              <Ripple
                className="css-8s1fw2 "
                color="#000"
                disabled={false}
                fill={true}
                onComplete={[Function]}
                overflow={false}
                size={null}
              >
                <div
                  aria-hidden={true}
                  className="css-8s1fw2 "
                  data-test-id="Ripple"
                  onClick={[Function]}
                  style={null}
                >
                  Text inside
                </div>
              </Ripple>
            </button>
          </Button>
        </Button>
      </RippleButton>
    </div>
  </ActionButton>
</ButtonLink>
`;
